### PR TITLE
Determine Homebrew Cellar from binary

### LINKED ISSUE #1653: OS X: "homebrew_packages" fails with non-standard Homebrew location
The `homebrew_packages` table is not populated with entries when *Homebrew* is not installed in the default location. E.g. I opted to separate my *Homebrew* installation into a dedicated subdirectory rather than placing it directly into the commonly recommended `/usr/local` location.

```
$ brew list | wc -l
     114
$ brew --cellar
/usr/local/homebrew/Cellar
```

```
osquery> .all homebrew_packages
osquery> 
```

It seems that osquery only checks the most common `Cellar` location (`/usr/local/Cellar`) rather than properly querying `brew` for the actual `Cellar` location (which might differ from `/usr/local/Cellar` depending on the user's installation/ environment). While osquery is running this can be verified with something like this:

```
$ sudo fs_usage | fgrep osquery
...
17:17:26  stat64            /usr/local/Cellar                                                                0.000007   osqueryi    
...
```

(That's not the path where my Homebrew `Cellar`is located, obviously. :smile: )

### PR DESCRIPTION
**Pull request by andrew-d**
Closes #1653

### CODE REVIEW

**Comment by osqueryer** [comment]
Can one of the osquery admins reply with "ok to test" to kick off a build...

File: osquery/tables/system/darwin/homebrew_packages.cpp [hunk 1]
```diff
@@ -56,10 +54,35 @@ std::vector<std::string> getHomebrewVersionsFromInfoPlistPath(
...
+  //   export HOMEBREW_BREW_FILE="$BREW_FILE_DIRECTORY/${0##*/}"
+  // Note that the `-P` flag to pwd resolves all symlinks.
+  //
+  // Next, it will use given filename to find the prefix:
+  //   HOMEBREW_PREFIX = Pathname.new(HOMEBREW_BREW_FILE).dirname.parent
+
+  // Get the actual location of the Homebrew binary.
+  // In the future, we could extend this to look at all 'brew' executables in
+  // $PATH and check all of them.
+  auto brewExecutable = fs::canonical("/usr/local/bin/brew");
```
Comment (sharvilshah) [inline osquery/tables/system/darwin/homebrew_packages.cpp]: Check if `/usr/local/bin/brew` exists.

**Comment by andrew-d** [comment]
@sharvilshah Good catch - thanks!  Added proper checks for path existence.  Is the usage of `boost::optional` here okay, or would you folks prefer some alternate method?

**Comment by sharvilshah** [comment]
Hey @andrew-d, I personally have no preference on `boost::optional`. But there is also [`osquery::Status`](https://github.com/facebook/osquery/blob/master/include/osquery/status.h) which is what rest of the codebase uses. One good thing about `osquery::Status` is that it's relatively straightforward to bubble up error string.

**Comment by andrew-d** [comment]
Done - switched to an `osquery::Status` style approach.

File: osquery/tables/system/darwin/homebrew_packages.cpp [hunk 2]
```diff
@@ -56,10 +56,57 @@ std::vector<std::string> getHomebrewVersionsFromInfoPlistPath(
...
+  // Note that the first `parent_path` call is to remove the filename, and the
+  // next to actually move up a directory.
+  prefix = brewExecutable.parent_path().parent_path();
+  return Status(0, "OK");
+}
+
+Status getHomebrewCellar(fs::path& cellarPath) {
+  fs::path path;
+
+  auto stat = getHomebrewPrefix(path);
```
Comment (theopolis) [inline osquery/tables/system/darwin/homebrew_packages.cpp]: No need to break this into two methods. Just find the Cellar and yield to an output argument in a single method.

File: osquery/tables/system/darwin/homebrew_packages.cpp (hunk 1 above)
Comment (theopolis) [inline osquery/tables/system/darwin/homebrew_packages.cpp]: Above the `kHomebrewBinary` symbol is used:
```cpp
kHomebrewBinary = "/usr/local/bin/brew";
```

Then below the static string is repeated. So it *is* checked for existence, but we're repeating const strings, think of the wasted bytes!!!! Oh the humanity!!!!!

File: osquery/tables/system/darwin/homebrew_packages.cpp [hunk 3]
```diff
@@ -56,10 +56,57 @@ std::vector<std::string> getHomebrewVersionsFromInfoPlistPath(
...
+  // directory that it is located in and concatenating `/../Library`:
+  //   BREW_FILE_DIRECTORY=$(chdir "${0%/*}" && pwd -P)
+  //   export HOMEBREW_BREW_FILE="$BREW_FILE_DIRECTORY/${0##*/}"
+  // Note that the `-P` flag to pwd resolves all symlinks.
+  //
+  // Next, it will use given filename to find the prefix:
+  //   HOMEBREW_PREFIX = Pathname.new(HOMEBREW_BREW_FILE).dirname.parent
+
+  if (!pathExists(kHomebrewBinary).ok()) {
+    return Status(1, "No Homebrew binary found");
```
Comment (theopolis) [inline osquery/tables/system/darwin/homebrew_packages.cpp]: nitpick, if you're not echoing or logging the status string/message then just return a `Status(1)` to not inflate the binary with unused const strings.

File: osquery/tables/system/darwin/homebrew_packages.cpp [hunk 4]
```diff
@@ -56,10 +56,57 @@ std::vector<std::string> getHomebrewVersionsFromInfoPlistPath(
...
+  fs::path path;
+
+  auto stat = getHomebrewPrefix(path);
+  if (!stat.ok()) {
+    return stat;
+  }
+
+  path /= "Cellar";
+  if (!pathExists(path).ok()) {
+    return Status(1, "No Homebrew Cellar found");
```
Comment (theopolis) [inline osquery/tables/system/darwin/homebrew_packages.cpp]: name nit as above for the status message, use it or lose it!

**Comment by theopolis** [comment]
Hey @andrew-d, this is looking good, another round of clearing up (or answering my comments with why I'm wrong) and we'll be ready to merge!

**Comment by andrew-d** [comment]
@theopolis Okay, addressed all review notes.  Once tests pass, I'll squash the commits into one :)

**Comment by theopolis** [comment]
ok to test

**Comment by osqueryer** [comment]
Build (5d91076148758d306d08b2cbacc3a25ff8a26c3f) is good! :+1:

**Comment by andrew-d** [comment]
@theopolis Tests passed, and I've rebased/force-pushed.

**Comment by osqueryer** [comment]
Build (5d91076148758d306d08b2cbacc3a25ff8a26c3f) is good! :+1:

**Pull request merged**